[BUG] DynamoDB connector can't handle array<bigint>

**Describe the bug**
I have a DynamoDB table that has the column `dataset_ids`, which is defined as a NumberSet. In the respective AWS Glue table schema that I've created for that dynamo table, I've defined that column as an `array<bigint>`. When I try to run the following query with Athena, I receive an error (stacktrace below):

```SELECT * FROM "dynamo-db-catalog"."default"."[redacted]" WHERE contains(dataset_ids, 104)```

**To Reproduce**
Steps to reproduce the behavior:
1. Create a dynamo table called `array_test`,  in which you have an item that has a key called `dataset_ids` with value `[104]`.
2. Create an AWS Glue schema for the DynamoDB table, defining `dataset_ids` column as an `array<bigint>`.
3. Query the table using athena: `SELECT * FROM "[catalog name]"."default"."athena_test" WHERE contains(dataset_ids, 104)`
4. See error

**Expected behavior**
Successful query.

**Screenshots / Exceptions / Errors**
Stack trace from CloudWatch:
```
2020-09-14 17:53:50 cb6c86aa-ebb4-4bd0-a669-024e9d8980df WARN  CompositeHandler:104 - handleRequest: Completed with an exception.
java.lang.RuntimeException: Error while processing field dataset_ids
	at com.amazonaws.athena.connectors.dynamodb.DynamoDBRecordHandler.lambda$readWithConstraint$1(DynamoDBRecordHandler.java:193) ~[task/:?]
	at com.amazonaws.athena.connector.lambda.data.S3BlockSpiller.writeRows(S3BlockSpiller.java:174) ~[task/:?]
	at com.amazonaws.athena.connectors.dynamodb.DynamoDBRecordHandler.readWithConstraint(DynamoDBRecordHandler.java:150) ~[task/:?]
	at com.amazonaws.athena.connector.lambda.handlers.RecordHandler.doReadRecords(RecordHandler.java:192) ~[task/:?]
	at com.amazonaws.athena.connector.lambda.handlers.RecordHandler.doHandleRequest(RecordHandler.java:158) ~[task/:?]
	at com.amazonaws.athena.connector.lambda.handlers.CompositeHandler.handleRequest(CompositeHandler.java:135) ~[task/:?]
	at com.amazonaws.athena.connector.lambda.handlers.CompositeHandler.handleRequest(CompositeHandler.java:100) [task/:?]
	at lambdainternal.EventHandlerLoader$2.call(EventHandlerLoader.java:902) [LambdaSandboxJava-1.0.jar:?]
	at lambdainternal.AWSLambda.startRuntime(AWSLambda.java:341) [LambdaSandboxJava-1.0.jar:?]
	at lambdainternal.AWSLambda.<clinit>(AWSLambda.java:63) [LambdaSandboxJava-1.0.jar:?]
	at java.lang.Class.forName0(Native Method) ~[?:1.8.0_201]
	at java.lang.Class.forName(Class.java:348) [?:1.8.0_201]
	at lambdainternal.LambdaRTEntry.main(LambdaRTEntry.java:150) [LambdaJavaRTEntry-1.0.jar:?]
Caused by: java.lang.RuntimeException: Unable to write value for field null_vector using value 239
	at com.amazonaws.athena.connector.lambda.data.BlockUtils.writeListValue(BlockUtils.java:859) ~[task/:?]
	at com.amazonaws.athena.connector.lambda.data.BlockUtils.writeList(BlockUtils.java:580) ~[task/:?]
	at com.amazonaws.athena.connector.lambda.data.BlockUtils.setComplexValue(BlockUtils.java:173) ~[task/:?]
	at com.amazonaws.athena.connector.lambda.data.Block.offerComplexValue(Block.java:218) ~[task/:?]
	at com.amazonaws.athena.connectors.dynamodb.DynamoDBRecordHandler.lambda$readWithConstraint$1(DynamoDBRecordHandler.java:172) ~[task/:?]
	... 12 more
Caused by: java.lang.ClassCastException: java.math.BigDecimal cannot be cast to java.lang.Long
	at com.amazonaws.athena.connector.lambda.data.BlockUtils.writeListValue(BlockUtils.java:793) ~[task/:?]
	at com.amazonaws.athena.connector.lambda.data.BlockUtils.writeList(BlockUtils.java:580) ~[task/:?]
	at com.amazonaws.athena.connector.lambda.data.BlockUtils.setComplexValue(BlockUtils.java:173) ~[task/:?]
	at com.amazonaws.athena.connector.lambda.data.Block.offerComplexValue(Block.java:218) ~[task/:?]
	at com.amazonaws.athena.connectors.dynamodb.DynamoDBRecordHandler.lambda$readWithConstraint$1(DynamoDBRecordHandler.java:172) ~[task/:?]
	... 12 more
```

**Connector Details (please complete the following information):**
 - Version: 2020.33.1
 - Name: serverlessrepo-AthenaDynamoDBConnector
 - Athena Query IDs
    - 232c5c4b-d9b1-41ee-b55d-180857879430.

**Additional context**
- Originally the column in AWS Glue was of type `set`, but I don't think that's supported at the moment.
- I've also tried setting the type to `array<int>` just to see if that would change anything, but I get the same error.

--- a/athena-dynamodb/src/main/java/com/amazonaws/athena/connectors/dynamodb/DynamoDBRecordHandler.java
+++ b/athena-dynamodb/src/main/java/com/amazonaws/athena/connectors/dynamodb/DynamoDBRecordHandler.java
@@ -24,10 +24,10 @@ import com.amazonaws.athena.connector.lambda.QueryStatusChecker;
 import com.amazonaws.athena.connector.lambda.ThrottlingInvoker;
 import com.amazonaws.athena.connector.lambda.data.Block;
 import com.amazonaws.athena.connector.lambda.data.BlockSpiller;
-import com.amazonaws.athena.connector.lambda.data.FieldResolver;
 import com.amazonaws.athena.connector.lambda.domain.Split;
 import com.amazonaws.athena.connector.lambda.handlers.RecordHandler;
 import com.amazonaws.athena.connector.lambda.records.ReadRecordsRequest;
+import com.amazonaws.athena.connectors.dynamodb.resolver.DynamoDBFieldResolver;
 import com.amazonaws.athena.connectors.dynamodb.util.DDBPredicateUtils;
 import com.amazonaws.athena.connectors.dynamodb.util.DDBRecordMetadata;
 import com.amazonaws.athena.connectors.dynamodb.util.DDBTypeUtils;
@@ -55,8 +55,6 @@ import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
 import java.io.IOException;
-import java.util.ArrayList;
-import java.util.Collection;
 import java.util.HashMap;
 import java.util.Iterator;
 import java.util.List;
@@ -139,6 +137,7 @@ public class DynamoDBRecordHandler
         invokerCache.get(tableName).setBlockSpiller(spiller);
         Iterator<Map<String, AttributeValue>> itemIterator = getIterator(split, tableName, recordsRequest.getSchema());
         DDBRecordMetadata recordMetadata = new DDBRecordMetadata(recordsRequest.getSchema());
+        DynamoDBFieldResolver resolver = new DynamoDBFieldResolver(recordMetadata);
         long numRows = 0;
         AtomicLong numResultRows = new AtomicLong(0);
         while (itemIterator.hasNext()) {
@@ -167,17 +166,18 @@ public class DynamoDBRecordHandler
                     try {
                         switch (fieldType) {
                             case LIST:
-                                // DDB may return Set so coerce to List
-                                List valueAsList = value != null ? new ArrayList((Collection) value) : null;
+                                // DDB may return Set so coerce to List. Also coerce each List item to the correct type.
+                                List valueAsList = value != null
+                                        ? DDBTypeUtils.coerceListToExpectedType(value, nextField, recordMetadata) : null;
                                 matched &= block.offerComplexValue(nextField.getName(),
                                         rowNum,
-                                        FieldResolver.DEFAULT,
+                                        resolver,
                                         valueAsList);
                                 break;
                             case STRUCT:
                                 matched &= block.offerComplexValue(nextField.getName(),
                                         rowNum,
-                                        (Field field, Object val) -> ((Map) val).get(field.getName()),
+                                        resolver,
                                         value);
                                 break;
                             default:

--- a/athena-dynamodb/src/main/java/com/amazonaws/athena/connectors/dynamodb/resolver/DynamoDBFieldResolver.java
+++ b/athena-dynamodb/src/main/java/com/amazonaws/athena/connectors/dynamodb/resolver/DynamoDBFieldResolver.java
@@ -1,0 +1,98 @@
+/*-
+ * #%L
+ * athena-dynamodb
+ * %%
+ * Copyright (C) 2019 - 2020 Amazon Web Services
+ * %%
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * 
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ * 
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * #L%
+ */
+package com.amazonaws.athena.connectors.dynamodb.resolver;
+
+import com.amazonaws.athena.connector.lambda.data.FieldResolver;
+import com.amazonaws.athena.connectors.dynamodb.util.DDBRecordMetadata;
+import com.amazonaws.athena.connectors.dynamodb.util.DDBTypeUtils;
+import org.apache.arrow.vector.types.Types;
+import org.apache.arrow.vector.types.pojo.Field;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Map;
+
+public class DynamoDBFieldResolver
+        implements FieldResolver
+{
+    private static final Logger logger = LoggerFactory.getLogger(DynamoDBFieldResolver.class);
+
+    final DDBRecordMetadata metadata;
+
+    public DynamoDBFieldResolver(DDBRecordMetadata recordMetadata)
+    {
+        metadata = recordMetadata;
+    }
+
+    /**
+     * Return the field value from a complex structure or list.
+     * @param field is the field that we would like to extract from the provided value.
+     * @param originalValue is the original value object.
+     * @return the field's value as a List for a LIST field type, a Map for a STRUCT field type, or the actual
+     * value if neither of the above.
+     * @throws IllegalArgumentException if originalValue is not an instance of Map.
+     * @throws RuntimeException if the fieldName does not exist in originalValue, or if the fieldType is a STRUCT and
+     * the fieldValue is not instance of Map.
+     */
+    @Override
+    public Object getFieldValue(Field field, Object originalValue)
+            throws RuntimeException
+    {
+        Types.MinorType fieldType = Types.getMinorTypeForArrowType(field.getType());
+        String fieldName = field.getName();
+        Object fieldValue;
+
+        if (originalValue instanceof Map) {
+            if (((Map) originalValue).containsKey(fieldName)) {
+                fieldValue = ((Map) originalValue).get(fieldName);
+            }
+            else {
+                throw new RuntimeException("Field not found in DB record: " + fieldName);
+            }
+        }
+        else {
+            throw new IllegalArgumentException("Invalid argument type. Expecting a Map, but got: " +
+                    originalValue.getClass().getTypeName());
+        }
+
+        switch (fieldType) {
+            case LIST:
+                return DDBTypeUtils.coerceListToExpectedType(fieldValue, field, metadata);
+            case STRUCT:
+                if (fieldValue instanceof Map) {
+                    // Both fieldType and fieldValue are nested structures => return as map.
+                    return fieldValue;
+                }
+                break;
+            default:
+                if (fieldValue instanceof Map) {
+                    // List of a Struct: array<struct<...>>
+                    return DDBTypeUtils.coerceValueToExpectedType(((Map<?, ?>) fieldValue).get(fieldName),
+                            field, fieldType, metadata);
+                }
+                else {
+                    return DDBTypeUtils.coerceValueToExpectedType(fieldValue, field, fieldType, metadata);
+                }
+        }
+
+        throw new RuntimeException("Invalid field value encountered in DB record for field: " + field +
+                ",value: " + fieldValue);
+    }
+}

--- a/athena-dynamodb/src/main/java/com/amazonaws/athena/connectors/dynamodb/resolver/DynamoDBFieldResolver.java
+++ b/athena-dynamodb/src/main/java/com/amazonaws/athena/connectors/dynamodb/resolver/DynamoDBFieldResolver.java
@@ -34,7 +34,7 @@ public class DynamoDBFieldResolver
 {
     private static final Logger logger = LoggerFactory.getLogger(DynamoDBFieldResolver.class);
 
-    final DDBRecordMetadata metadata;
+    private final DDBRecordMetadata metadata;
 
     public DynamoDBFieldResolver(DDBRecordMetadata recordMetadata)
     {

--- a/athena-dynamodb/src/main/java/com/amazonaws/athena/connectors/dynamodb/util/DDBTypeUtils.java
+++ b/athena-dynamodb/src/main/java/com/amazonaws/athena/connectors/dynamodb/util/DDBTypeUtils.java
@@ -320,8 +320,10 @@ public final class DDBTypeUtils
      * @param field is the LIST field containing a list type in the child field.
      * @param recordMetadata contains metadata information.
      * @return a List of coerced values.
+     * @throws RuntimeException when value is instance of Map since a List is expected.
      */
-    public static List coerceListToExpectedType(Object value, Field field, DDBRecordMetadata recordMetadata)
+    public static List<Object> coerceListToExpectedType(Object value, Field field, DDBRecordMetadata recordMetadata)
+            throws RuntimeException
     {
         Field childField = field.getChildren().get(0);
         Types.MinorType fieldType = Types.getMinorTypeForArrowType(childField.getType());

--- a/athena-dynamodb/src/main/java/com/amazonaws/athena/connectors/dynamodb/util/DDBTypeUtils.java
+++ b/athena-dynamodb/src/main/java/com/amazonaws/athena/connectors/dynamodb/util/DDBTypeUtils.java
@@ -245,19 +245,29 @@ public final class DDBTypeUtils
             }
             value = coerceDateTimeToExpectedType(value, fieldType, dateTimeFormat, recordMetadata.getDefaultTimeZone());
         }
-        else if (!fieldType.equals(Types.MinorType.DECIMAL) && value instanceof BigDecimal) {
-            value = DDBTypeUtils.coerceDecimalToExpectedType((BigDecimal) value, fieldType);
+        else if (value instanceof Number) {
+            // Number conversion from any DDB Numeric type to the correct field numeric type (as specified in schema)
+            value = DDBTypeUtils.coerceNumberToExpectedType((Number) value, fieldType);
         }
         return value;
     }
 
-    private static Object coerceDecimalToExpectedType(BigDecimal value, Types.MinorType fieldType)
+    /**
+     * Converts a numeric value extracted from the DDB record to the correct type (Integer, Long, etc..) expected
+     * by Arrow for the field type.
+     * @param value is the number extracted from the DDB record.
+     * @param fieldType is the Arrow MinorType.
+     * @return the converted value.
+     */
+    private static Object coerceNumberToExpectedType(Number value, Types.MinorType fieldType)
     {
         switch (fieldType) {
             case INT:
-            case TINYINT:
-            case SMALLINT:
                 return value.intValue();
+            case TINYINT:
+                return value.byteValue();
+            case SMALLINT:
+                return value.shortValue();
             case BIGINT:
                 return value.longValue();
             case FLOAT4:
@@ -301,5 +311,38 @@ public final class DDBTypeUtils
             ex.printStackTrace();
             return value;
         }
+    }
+
+    /**
+     * Converts a Set to a List, and coerces all list items into the correct type. If value is not a Collection, this
+     * method will return a List containing a single item.
+     * @param value is the Set/List of items.
+     * @param field is the LIST field containing a list type in the child field.
+     * @param recordMetadata contains metadata information.
+     * @return a List of coerced values.
+     */
+    public static List coerceListToExpectedType(Object value, Field field, DDBRecordMetadata recordMetadata)
+    {
+        Field childField = field.getChildren().get(0);
+        Types.MinorType fieldType = Types.getMinorTypeForArrowType(childField.getType());
+
+        if (!(value instanceof Collection)) {
+            if (value instanceof Map) {
+                throw new RuntimeException("Unexpected type (Map) encountered for: " + childField.getName());
+            }
+            return Collections.singletonList(coerceValueToExpectedType(value, childField, fieldType, recordMetadata));
+        }
+
+        List<Object> coercedList = new ArrayList<>();
+        if (fieldType == Types.MinorType.LIST) {
+            // Nested lists: array<array<...>, array<...>, ...>
+            ((Collection<?>) value).forEach(list -> coercedList
+                    .add(coerceListToExpectedType(list, childField, recordMetadata)));
+        }
+        else {
+            ((Collection<?>) value).forEach(item -> coercedList
+                    .add(coerceValueToExpectedType(item, childField, fieldType, recordMetadata)));
+        }
+        return coercedList;
     }
 }

--- a/athena-dynamodb/src/test/java/com/amazonaws/athena/connectors/dynamodb/DynamoDBMetadataHandlerTest.java
+++ b/athena-dynamodb/src/test/java/com/amazonaws/athena/connectors/dynamodb/DynamoDBMetadataHandlerTest.java
@@ -244,7 +244,7 @@ public class DynamoDBMetadataHandlerTest
 
         assertThat(res.getTableName().getSchemaName(), equalTo(DEFAULT_SCHEMA));
         assertThat(res.getTableName().getTableName(), equalTo(TEST_TABLE));
-        assertThat(res.getSchema().getFields().size(), equalTo(10));
+        assertThat(res.getSchema().getFields().size(), equalTo(11));
     }
 
     @Test

--- a/athena-dynamodb/src/test/java/com/amazonaws/athena/connectors/dynamodb/TestBase.java
+++ b/athena-dynamodb/src/test/java/com/amazonaws/athena/connectors/dynamodb/TestBase.java
@@ -136,6 +136,8 @@ public class TestBase
             item.put("col_8", toAttributeValue(ImmutableList.of(ImmutableMap.of("mostlyEmptyMap",
                     i % 128 == 0 ? ImmutableMap.of("subtractions", ImmutableSet.of(i - 100, i - 200)) : ImmutableMap.of()))));
             item.put("col_9", toAttributeValue(100.0f + i));
+            item.put("col_10", toAttributeValue(ImmutableList.of(ImmutableList.of(1 * i, 2 * i, 3 * i),
+                    ImmutableList.of(4 * i, 5 * i), ImmutableList.of(6 * i, 7 * i, 8 * i))));
             tableWriteItems.addItemToPut(toItem(item));
 
             if (tableWriteItems.getItemsToPut().size() == 25) {

